lab3/client/audio_processing.py

```diff
@@ -0,0 +1,276 @@
+            raise ValueError
+
+    def call_api(self, url):
+        response = requests.post(
+            url,
+            files={"file": (self.file_name, self.file_bytes, self.file_mime_type)},
+            timeout=8000,
+        )
+        if response.status_code != 200:
+            st.error(f"error {response.status_code} {response.content}")
+        return response.json()
+
+    def convert_audio_bytes(self, file_bytes: Optional[bytes] = None, file_mime_type: Optional[str] = None):
+        if file_bytes is None:
+            file_bytes = self.file_bytes
+        if file_mime_type is None:
+            file_mime_type = self.file_mime_type
+
+        buffer = io.BytesIO()
+        buffer.write(file_bytes)
+        buffer.seek(0)
+
+        audio = getattr(pydub.AudioSegment, self.mime_type_audio_command_map[file_mime_type])(file=buffer)
+        # handle audio data
+        channel_sounds = audio.split_to_mono()
+        samples = [s.get_array_of_samples() for s in channel_sounds]
+        fp_arr = np.array(samples).T.astype(np.float32)
+        fp_arr /= np.iinfo(samples[0].typecode).max
+        source, sample_rate = fp_arr[:, 0], audio.frame_rate
+        source = resampy.resample(source, sample_rate, SAMPLE_RATE, axis=0, filter="kaiser_best")
+        return source, SAMPLE_RATE
+
+    @staticmethod
+    def draw_audio_player(source, sample_rate):
+        in_memory_file = io.BytesIO()
+        wavfile.write(in_memory_file, rate=sample_rate, data=source)
+        # draw player
+        st.audio(in_memory_file)
+
+    @staticmethod
+    def plot_wave(source, sample_rate):
+        fig, ax = plt.subplots()
+        librosa.display.waveshow(source, sr=sample_rate, x_axis="time", ax=ax)
+        return plt.gcf()
+
+    @staticmethod
+    def add_h_space():
+        st.markdown("<br></br>", unsafe_allow_html=True)
+
+
+class SpeechRecognition(AbstractOption):
+    name = "Speech Recognition"
+
+    def handle(self, *args, **kwargs):
+
+        api_result = self.call_api(endpoint_recognition)
+
+        st.markdown(
+            "<h4 style='text-align: center; color: black;'>Audio</h5>",
+            unsafe_allow_html=True,
+        )
+        # draw audio player
+        source, sample_rate = self.convert_audio_bytes()
+        self.draw_audio_player(source, sample_rate)
+        st.markdown("---")
+        # show recognition result
+        st.write(api_result["text"])
+
+
+class SpeechSeparation(AbstractOption):
+    name = "Speech Separation"
+
+    def handle(self, *args, **kwargs):
+        # start recognition via api
+        api_result = self.call_api(endpoint_separation)
+        separated_file = api_result.get("output_files")
+        separated_file.sort(key=lambda x: x.get("order"))
+        for i, source in enumerate(separated_file):
+            result = source["file"]
+            result = result.encode(encoding="UTF-8")
+            buff = base64.decodebytes(result)
+            sound = np.frombuffer(buff, dtype=np.float32)
```

Comment:
Дублирование кода предлагаю вынести в отдельную функцию, например encode2numpy
```suggestion
            def encode2numpy(base64str):
                  result = base64str.encode(encoding="UTF-8")
                  buff = base64.decodebytes(base64str)
                  sound = np.frombuffer(buff, dtype=np.float32)
```

Reply:
не вижу особого выйгрыша. 4 строчки повторяются в двух местах. места потратится +- столько же, только еще придется платить за вызов функции.